### PR TITLE
Add clientset generator

### PR DESCRIPTION
**Pull request by mfojtik**
TODO:

- [x] Add `verify` script for CI to check if the generated clients matches with PR clients
- [x] Fix the client names for things like SDN, OAuth, etc..
- [x] Add versions to generated clients (12, 13, ...)
- [x] Generate fake clients for testing

Fixes: https://github.com/openshift/origin/issues/9369

### CODE REVIEW

**Comment by mfojtik** [comment]
@deads2k @soltysh @smarterclayton FYI

**Comment by deads2k** [comment]
I like the idea of one clientset per eventual group that we'll have.  I think it will encourage people to get dependencies correct and make something like the docker registry split a little cleaner.

File: pkg/authorization/client/clientset_generated/internalclientset/clientset.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/golang/glog"
+	v1core "github.com/openshift/origin/pkg/authorization/client/clientset_generated/internalclientset/typed/core/v1"
+	restclient "k8s.io/kubernetes/pkg/client/restclient"
+	discovery "k8s.io/kubernetes/pkg/client/typed/discovery"
+	"k8s.io/kubernetes/pkg/util/flowcontrol"
+)
+
+type Interface interface {
+	Discovery() discovery.DiscoveryInterface
+	Core() v1core.CoreInterface
```
Comment (deads2k) [inline pkg/authorization/client/clientset_generated/internalclientset/clientset.go]: This looks strange to me.  `Core`?

File: pkg/authorization/client/clientset_generated/internalclientset/clientset.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import (
+	"github.com/golang/glog"
+	v1core "github.com/openshift/origin/pkg/authorization/client/clientset_generated/internalclientset/typed/core/v1"
+	restclient "k8s.io/kubernetes/pkg/client/restclient"
+	discovery "k8s.io/kubernetes/pkg/client/typed/discovery"
+	"k8s.io/kubernetes/pkg/util/flowcontrol"
+)
+
+type Interface interface {
+	Discovery() discovery.DiscoveryInterface
```
Comment (deads2k) [inline pkg/authorization/client/clientset_generated/internalclientset/clientset.go]: Can this actually discover the correct version, since we're pointed at `oapi`?

File: pkg/authorization/client/clientset_generated/internalclientset/clientset.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/authorization/client/clientset_generated/internalclientset/clientset.go]: It's effectively the inner interface, and was convention "decided" upstream.

On Jun 16, 2016, at 9:09 AM, David Eads <notifications@github.com> wrote:

In
pkg/authorization/client/clientset_generated/internalclientset/clientset.go
<https://github.com/openshift/origin/pull/9372#discussion_r67340417>:

> @@ -0,0 +1,73 @@
> +package internalclientset
> +
> +import (
> +	"github.com/golang/glog"
> +	v1core "github.com/openshift/origin/pkg/authorization/client/clientset_generated/internalclientset/typed/core/v1"
> +	restclient "k8s.io/kubernetes/pkg/client/restclient"
> +	discovery "k8s.io/kubernetes/pkg/client/typed/discovery"
> +	"k8s.io/kubernetes/pkg/util/flowcontrol"
> +)
> +
> +type Interface interface {
> +	Discovery() discovery.DiscoveryInterface
> +	Core() v1core.CoreInterface

This looks strange to me. Core?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/9372/files/2a0ceea3c7e61fc7e21b9e5471a651781638e7d5#r67340417>,
or mute the thread
<https://github.com/notifications/unsubscribe/ABG_p9QklkgM94JMGs_2obbASJ_0FiWzks5qMUsRgaJpZM4I3W73>
.

**Comment by mfojtik** [comment]
@smarterclayton i guess the output dir will need the version

**Comment by mfojtik** [comment]
and we also need to generate fake clients

File: pkg/authorization/client/clientset_generated/internalclientset/clientset.go (hunk 2 above)
Comment (mfojtik) [inline pkg/authorization/client/clientset_generated/internalclientset/clientset.go]: If not, we will need to fix that in upstream generator?

**Comment by smarterclayton** [comment]
We should settle now on our groups:

builds.openshift.io
deployments.
routes.
authorization
users
security

Basically the pkg name but made into a full word.

On Jun 16, 2016, at 9:09 AM, David Eads <notifications@github.com> wrote:

I like the idea of one clientset per eventual group that we'll have. I
think it will encourage people to get dependencies correct and make
something like the docker registry split a little cleaner.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/9372#issuecomment-226479898>, or mute
the thread
<https://github.com/notifications/unsubscribe/ABG_p3TcCxYJzF7-JBrnWx9p39EcAjIeks5qMUr_gaJpZM4I3W73>
.

**Comment by mfojtik** [comment]
@smarterclayton templates.openshift.io and networking.openshift.io ?

**Comment by deads2k** [comment]
plural or singular?  I don't feel strongly either way, but I want to make an explicit choice.  We currently have these packages:

 1. authorization
 1. build
 1. deploy
 1. image
 1. oauth
 1. project
 1. route
 1. sdn - networking?
 1. template
 1. user
 1. security - I don't like the name.  This holds SCC today.  

I'm ok with most of them, I called out the ones I don't like.  I think we should also figure out how to add the configapi as a group.

**Comment by mfojtik** [comment]
@deads2k I like 'networking' (we use full names except oauth ;-)

**Comment by smarterclayton** [comment]
Does upstream have pluralization?  If no singular is fine.  SDN should be
networking.

What would you suggest besides SCC?

**Comment by smarterclayton** [comment]
server.config.openshift.io

**Comment by smarterclayton** [comment]
oauth is technically a full name so we're covered there.

**Comment by deads2k** [comment]
> Does upstream have pluralization?

Upstream is inconsistent. "batch" versus "extensions" and "metrics"

**Comment by soltysh** [comment]
> Upstream is inconsistent. "batch" versus "extensions" and "metrics"

There are "apps" but "autoscaling". I vote for singular, since majority of groups upstream is singular.

**Comment by soltysh** [comment]
> server.config.openshift.io 

That's too broad, imho. Security might be better.

>oauth is technically a full name so we're covered there.

Instead of authorization and oauth, I'd suggest using authz and authn, or full names (authorization and authentication).

**Comment by deads2k** [comment]
> Instead of authorization and oauth, I'd suggest using authz and authn, or full names (authorization and authentication).

authentication would cover oauth and user?  The two aren't closely related, but aren't the same.  @liggitt opinion?

**Comment by mfojtik** [comment]
@soltysh I would prefer avoiding short names as 'authz' if possible.

**Comment by smarterclayton** [comment]
Why wouldn't all our server configuration types be under the same group?

These are not authentication APIs, they are literally Oauth apis.  No
generic authentication in them.  No short forms either.

File: pkg/authorization/api/types.go [hunk 3]
```diff
@@ -105,6 +105,7 @@ type RoleBinding struct {
 	RoleRef kapi.ObjectReference
 }
 
+// +genclient=true
```
Comment (rhcarvalho) [inline pkg/authorization/api/types.go]: Won't this pollute the godoc output?

Comment (mfojtik) [inline pkg/authorization/api/types.go]: negative, but I should put extra new line after... similar to upstream: https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L344

File: pkg/authorization/client/clientset_generated/release_1_3/typed/core/v1/core_client.go [hunk 4]
```diff
@@ -0,0 +1 @@
+package v1
```
Comment (mfojtik) [inline pkg/authorization/client/clientset_generated/release_1_3/typed/core/v1/core_client.go]: @smarterclayton I wonder why is this empty :/

Comment (mfojtik) [inline pkg/authorization/client/clientset_generated/release_1_3/typed/core/v1/core_client.go]: nvmd. fixed.

**Comment by mfojtik** [comment]
@smarterclayton if it holds SCC, we can call it 'scc' or 'security-constrains'

**Comment by mfojtik** [comment]
Added group names, I will change them if we decide to use different names.

File: pkg/authorization/api/register.go [hunk 5]
```diff
@@ -5,7 +5,7 @@ import (
 	"k8s.io/kubernetes/pkg/runtime"
 )
 
-const GroupName = ""
+const GroupName = "authorization.openshift.io"
```
Comment (deads2k) [inline pkg/authorization/api/register.go]: Not just yet.  The master can't handle it https://github.com/kubernetes/kubernetes/pull/27554 yet.

File: pkg/api/register.go [hunk 6]
```diff
@@ -17,7 +17,7 @@ import (
 	_ "github.com/openshift/origin/pkg/user/api"
 )
 
-const GroupName = ""
+const GroupName = "openshift.io"
```
Comment (deads2k) [inline pkg/api/register.go]: I'm not sure we'll ever want this one set up that way.

File: pkg/cmd/server/api/v1/register.go [hunk 7]
```diff
@@ -5,7 +5,7 @@ import (
 	"k8s.io/kubernetes/pkg/runtime"
 )
 
-const GroupName = ""
+const GroupName = "server.openshift.io"
```
Comment (deads2k) [inline pkg/cmd/server/api/v1/register.go]: This would break deserializing existing master-configs on update.

Comment (mfojtik) [inline pkg/cmd/server/api/v1/register.go]: k, will remove this

File: pkg/authorization/api/register.go (hunk 5 above)
Comment (mfojtik) [inline pkg/authorization/api/register.go]: @deads2k so we need the kube PR to land first on order to be able to merge this?

Comment (deads2k) [inline pkg/authorization/api/register.go]: > @deads2k so we need the kube PR to land first on order to be able to merge this?

Can we generate the clients using the `""` group for now, then generate another set (we'll need both anway) after we register with both names.  

We divvied up groups so we'd know whether any shouldn't be generated single folder.

**Comment by mfojtik** [comment]
@deads2k removed the group names for now, PTAL

@smarterclayton is there a way to pick the name for the clientset other than having the go comment on top of the "field" in types.go where that field name will be used as a name? I added "Security" struct just to have name "security" ;-)

File: hack/update-generated-clientsets.sh [hunk 8]
```diff
@@ -0,0 +1,62 @@
+#!/bin/bash
```
Comment (deads2k) [inline hack/update-generated-clientsets.sh]: @stevekuznetsov have a look.

**Comment by deads2k** [comment]
Where are the unversioned packages?  I'd expect those to be the ones we'd actually use for the most part.

File: pkg/security/api/types.go [hunk 9]
```diff
@@ -5,6 +5,11 @@ import (
 	"k8s.io/kubernetes/pkg/api/unversioned"
 )
 
+// +genclient=true
+
+// Security is just a name for the client generated
+type Security struct{}
```
Comment (deads2k) [inline pkg/security/api/types.go]: This seems to generate a client that will always fail.  What is the intent for this?

File: hack/update-generated-clientsets.sh [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    echo "It looks as if you don't have a compiled client-gen binary"
+    echo
+    echo "If you are running from a clone of the git repo, please run"
+    echo "'./hack/build-go.sh Godeps/_workspace/src/k8s.io/kubernetes/cmd/libs/go2idl/client-gen'."
+  } >&2
+  exit 1
+fi
+
+# Acquire the version string used as a suffix for the client_set
+origin_version=$(cd $OS_ROOT && git describe --abbrev=0 --tags | sed -e "s/^v\([0-9]*\)\.\([0-9]*\).*/\1_\2/")
```
Comment (stevekuznetsov) [inline hack/update-generated-clientsets.sh]: I don't think regex this complicated (with named groups, of all things) really is maintainable. Do we know the output of `git describe` doesn't change? Can we get this info some other way? If this is some sort of version we're exposing externally, should it be exposed in `oc version` or similar?

File: hack/update-generated-clientsets.sh [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  github.com/openshift/origin/pkg/security
+)
+
+
+function generate_clientset_for() {
+  local package="$1";shift
+  local name="$1";shift
+  pushd ${OS_ROOT} >/dev/null
+  local common_args="--go-header-file hack/boilerplate.txt --input=api/v1"
+  $clientgen --clientset-path $package/client/clientset_generated \
```
Comment (stevekuznetsov) [inline hack/update-generated-clientsets.sh]: Be careful with the way strings break on spaces ... 

```
+  $clientgen --clientset-path="${package}/client/clientset_generated" \
+             --input-base="${package}/api"                            \
+             --clientset-name="${name}"                               \
+             ${common_args}                                           \
+             "$@"
```

File: hack/update-generated-clientsets.sh [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+  github.com/openshift/origin/pkg/user
+  github.com/openshift/origin/pkg/security
+)
+
+
+function generate_clientset_for() {
+  local package="$1";shift
+  local name="$1";shift
+  pushd ${OS_ROOT} >/dev/null
+  local common_args="--go-header-file hack/boilerplate.txt --input=api/v1"
```
Comment (stevekuznetsov) [inline hack/update-generated-clientsets.sh]: Should be `--go-header-file=hack/boilerplate.txt` for consistency

File: hack/update-generated-clientsets.sh [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+  $clientgen --clientset-path $package/client/clientset_generated \
+    --input-base=$package/api --clientset-name=$name $common_args $@
+  popd >/dev/null
+}
+
+for pkg in "${packages[@]}"; do
+  if [[ "${1:-''}" != "--verify-only" ]]; then
```
Comment (stevekuznetsov) [inline hack/update-generated-clientsets.sh]: We don't need another method of doing this -- if I remember correctly I used `VERIFY=true hack/update...` for the swagger documentation script. In either case, we should prefer envars for this.

File: hack/update-generated-clientsets.sh [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    # Remove deprecated/old files
+    rm -rf ${GOPATH}/src/${pkg}/client/clientset_generated
```
Comment (stevekuznetsov) [inline hack/update-generated-clientsets.sh]: Needs quotes for correct breaking on spaces and a check for the first envar so we don't attempt to `rm -rf /src...`
```
rm -rf "${GOPATH:?}/src/${pkg}/client/clientset_generated"
```

File: hack/update-generated-clientsets.sh [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+  fi
+
+  generate_clientset_for $pkg "internalclientset" "$@"
```
Comment (stevekuznetsov) [inline hack/update-generated-clientsets.sh]: Quote `"${pkg}"` here to ensure it's a single arg

File: hack/update-generated-clientsets.sh (hunk 14 above)
Comment (rhcarvalho) [inline hack/update-generated-clientsets.sh]: Could use `go list -f '{{.Dir}}' "${pkg}"` to get the directory, that would work also for people with multiple paths in GOPATH (a perfectly valid configuration).

File: hack/update-generated-clientsets.sh (hunk 10 above)
Comment (mfojtik) [inline hack/update-generated-clientsets.sh]: @stevekuznetsov kube use even more complicated regexp ;-)

see: https://github.com/kubernetes/kubernetes/blob/master/hack/lib/version.sh#L59

File: hack/update-generated-clientsets.sh (hunk 13 above)
Comment (mfojtik) [inline hack/update-generated-clientsets.sh]: fixed.

File: hack/update-generated-clientsets.sh (hunk 14 above)
Comment (mfojtik) [inline hack/update-generated-clientsets.sh]: @rhcarvalho thanks! i will use that over rm.

**Comment by mfojtik** [comment]
@deads2k  `pkg/build/client/clientset_generated/internalclientset` ?

File: pkg/security/api/types.go (hunk 9 above)
Comment (mfojtik) [inline pkg/security/api/types.go]: what you mean it will always fail? we don't want to generate client for security ?

File: hack/update-generated-clientsets.sh (hunk 10 above)
Comment (soltysh) [inline hack/update-generated-clientsets.sh]: Can't we just use `os::build::get_version_vars` and replace dots with underscores?

File: hack/update-generated-clientsets.sh [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+  github.com/openshift/origin/pkg/build
+  github.com/openshift/origin/pkg/deploy
+  github.com/openshift/origin/pkg/image
+  github.com/openshift/origin/pkg/oauth
+  github.com/openshift/origin/pkg/project
+  github.com/openshift/origin/pkg/route
+  github.com/openshift/origin/pkg/sdn
+  github.com/openshift/origin/pkg/template
+  github.com/openshift/origin/pkg/user
+  github.com/openshift/origin/pkg/security
```
Comment (soltysh) [inline hack/update-generated-clientsets.sh]: Make them in alphabetic order for better searchability.

File: hack/update-generated-clientsets.sh [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+)
+
+
+function generate_clientset_for() {
+  local package="$1";shift
+  local name="$1";shift
+  pushd ${OS_ROOT} >/dev/null
+  local common_args="--go-header-file=hack/boilerplate.txt"
```
Comment (soltysh) [inline hack/update-generated-clientsets.sh]: Where is that file? I don't see it being part of this commit?

File: Makefile [hunk 18]
```diff
@@ -53,6 +53,7 @@ verify: build
 	hack/verify-govet.sh
 	hack/verify-generated-deep-copies.sh
 	hack/verify-generated-conversions.sh
+	hack/verify-generated-clientsets.sh
```
Comment (soltysh) [inline Makefile]: You need to add `hack/update-generated-sclientsets.sh` to `update` target as well.

File: pkg/authorization/api/types.go (hunk 3 above)
Comment (soltysh) [inline pkg/authorization/api/types.go]: Make sure to check all the types are generated correctly. Although the tag should work one for a file, for some reasons for ScheduledJobs (while adding them to k8s) I had to specify once again the tag right before ScheduledJobs types in internal types.go.

File: hack/verify-generated-clientsets.sh [hunk 19]
```diff
@@ -8,7 +8,7 @@ source "${OS_ROOT}/hack/lib/init.sh"
 
 cd "${OS_ROOT}"
 echo "===== Verifying Generated Client sets ====="
-output=$(VERIFY=true ${OS_ROOT}/hack/update-generated-clientsets.sh 2>&1)
+output=$(VERIFY=true ${OS_ROOT}/hack/update-generated-clientsets.sh)
```
Comment (soltysh) [inline hack/verify-generated-clientsets.sh]: Move this change to the first commit, where you're adding that script.

File: hack/update-generated-clientsets.sh [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+  $clientgen --clientset-path="${package}/client/clientset_generated" \
```
Comment (stevekuznetsov) [inline hack/update-generated-clientsets.sh]: Please format this better, ex how I had in my previous comment

File: hack/update-generated-clientsets.sh [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+    --input-base="${package}/api" --clientset-name="${name}" $common_args "$@"
+  popd >/dev/null
+}
+
+verify="${VERIFY:-}"
+
+for pkg in "${packages[@]}"; do
+  if [[ ! -n "${verify}" ]]; then
```
Comment (stevekuznetsov) [inline hack/update-generated-clientsets.sh]: Opposite of `-n` is `-z`

File: hack/update-generated-clientsets.sh (hunk 10 above)
Comment (stevekuznetsov) [inline hack/update-generated-clientsets.sh]: Isn't there code in `oc version` doing something very similar? If not, let's have named variables giving us each part and then explicit concatenation. One-liners won't help anyone understand or maintain this.

File: hack/verify-generated-clientsets.sh (hunk 19 above)
Comment (stevekuznetsov) [inline hack/verify-generated-clientsets.sh]: Also, why is this necessary? Why do we not want to capture all output?

File: hack/update-generated-clientsets.sh (hunk 10 above)
Comment (smarterclayton) [inline hack/update-generated-clientsets.sh]: We have a build time helper for this os::build::git something.

File: hack/update-generated-clientsets.sh (hunk 17 above)
Comment (mfojtik) [inline hack/update-generated-clientsets.sh]: @soltysh already in hack/ (I think it is empty, conversion generator use this as well)

File: hack/verify-generated-clientsets.sh (hunk 19 above)
Comment (mfojtik) [inline hack/verify-generated-clientsets.sh]: we do, this was just debugging, because "sometimes" I get error 255 in travis....

File: hack/update-generated-clientsets.sh (hunk 20 above)
Comment (mfojtik) [inline hack/update-generated-clientsets.sh]: done.

File: Makefile (hunk 18 above)
Comment (mfojtik) [inline Makefile]: done.

File: hack/update-generated-clientsets.sh (hunk 10 above)
Comment (soltysh) [inline hack/update-generated-clientsets.sh]: This is the remaining piece for me to accept this PR.

**Comment by mfojtik** [comment]
@deads2k we are green now.

**Comment by soltysh** [comment]
Whereas I'm gonna keep on complaining about that grep magic to get origin version, please use one of the functions we already have.

File: hack/update-generated-clientsets.sh (hunk 10 above)
Comment (mfojtik) [inline hack/update-generated-clientsets.sh]: fixed.

**Comment by mfojtik** [comment]
@soltysh fixed

**Comment by soltysh** [comment]
LGTM

**Comment by deads2k** [comment]
lgtm, we can build out what we need from here.

Anyone else want a look before merge?

**Comment by soltysh** [comment]
@smarterclayton / @liggitt  any last comments or we're good to go with this?

File: pkg/route/api/v1/types.go [hunk 22]
```diff
@@ -6,6 +6,7 @@ import (
 	"k8s.io/kubernetes/pkg/util/intstr"
 )
 
+// +genclient=true
```
Comment (rhcarvalho) [inline pkg/route/api/v1/types.go]: missing new line here to don't mix with godoc.

File: pkg/authorization/client/clientset_generated/internalclientset/clientset.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/authorization/client/clientset_generated/internalclientset/clientset.go]: Would like the answer to this.

**Comment by smarterclayton** [comment]
I had a comment on the Discovery questions from David.

File: pkg/route/api/v1/types.go (hunk 22 above)
Comment (mfojtik) [inline pkg/route/api/v1/types.go]: fixed.

File: pkg/authorization/client/clientset_generated/internalclientset/clientset.go (hunk 2 above)
Comment (mfojtik) [inline pkg/authorization/client/clientset_generated/internalclientset/clientset.go]: I don't think this work, because https://github.com/kubernetes/kubernetes/blob/cbab337b9967663f581165da18126f660b7fa7a8/pkg/client/typed/discovery/discovery_client.go#L291

I think we will have to make the `LegacyPrefix` configurable?

Comment (mfojtik) [inline pkg/authorization/client/clientset_generated/internalclientset/clientset.go]: @smarterclayton will public `var LegacyPrefix` do or we can do smarter plumbing?

Comment (smarterclayton) [inline pkg/authorization/client/clientset_generated/internalclientset/clientset.go]: We can potentially punt on discovery via this interface for now.

Comment (mfojtik) [inline pkg/authorization/client/clientset_generated/internalclientset/clientset.go]: @smarterclayton I can spawn a follow up issue to fix Discovery

**Comment by deads2k** [comment]
[merge], but @mfojtik promises to open an issue and deal with discovery.

**Comment by mfojtik** [comment]
[merge]

**Comment by mfojtik** [comment]
[merge]
